ledger/nomos-ledger/src/mantle/sdp/rewards/blend/target_session.rs

```diff
@@ -0,0 +1,227 @@
+use std::{cmp::Ordering, collections::HashMap, iter::once};
+
+use nomos_blend_message::reward::{BlendingTokenEvaluation, HammingDistance};
+use nomos_blend_proofs::{quota::VerifiedProofOfQuota, selection::VerifiedProofOfSelection};
+use nomos_core::{
+    mantle::Utxo,
+    sdp::{ProviderId, ServiceType, SessionNumber},
+};
+use rpds::{HashTrieMapSync, HashTrieSetSync};
+use zksign::PublicKey;
+
+use crate::mantle::sdp::rewards::{
+    Error,
+    blend::{RewardsParameters, current_session::CurrentSessionState},
+    distribute_rewards,
+};
+
+/// The immutable state of the target session for which rewards are being
+/// calculated. The target session is `s-1` if `s` is the current session.
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct TargetSessionState {
+    /// The target session number
+    session_number: SessionNumber,
+    /// The providers in the target session
+    providers: HashTrieMapSync<ProviderId, PublicKey>,
+    /// Parameters for evaluating activity proofs in the target session
+    token_evaluation: BlendingTokenEvaluation,
+    // TODO: Add proof verifiers
+}
+
+impl TargetSessionState {
+    pub const fn new(
+        session_number: SessionNumber,
+        providers: HashTrieMapSync<ProviderId, PublicKey>,
+        token_evaluation: BlendingTokenEvaluation,
+    ) -> Self {
+        Self {
+            session_number,
+            providers,
+            token_evaluation,
+        }
+    }
+
+    pub const fn session_number(&self) -> SessionNumber {
+        self.session_number
+    }
+
+    pub fn providers(&self) -> impl Iterator<Item = (&ProviderId, &PublicKey)> {
+        self.providers.iter()
+    }
+
+    fn num_providers(&self) -> u64 {
+        self.providers
+            .size()
+            .try_into()
+            .expect("number of providers must fit in u64")
+    }
+
+    pub fn verify_proof(
+        &self,
+        provider_id: &ProviderId,
+        proof: &nomos_core::sdp::blend::ActivityProof,
+        current_session_state: &CurrentSessionState,
+        settings: &RewardsParameters,
+    ) -> Result<(PublicKey, HammingDistance), Error> {
+        if proof.session != self.session_number {
+            return Err(Error::InvalidSession {
+                expected: self.session_number,
+                got: proof.session,
+            });
+        }
+
+        let num_providers = self.num_providers();
+        assert!(
+            num_providers >= settings.minimum_network_size.get(),
+            "number of providers must be >= minimum_network_size"
+        );
+        let &zk_id = self
+            .providers
+            .get(provider_id)
+            .ok_or_else(|| Error::UnknownProvider(Box::new(*provider_id)))?;
+
+        let proof = verify_activity_proof(proof);
```

Comment:
I guess this should be call `verified_proof`? I was going to suggest we make use of the typesystem, but realized input and output types are already different, so maybe just naming it better would be less confusing.

Reply:
resolved [fca6e72](https://github.com/logos-blockchain/nomos/pull/1967/commits/fca6e728310eae55052621dc2fe2695f5718a40e)